netket/utils/static_range.py

```diff
@@ -100,18 +106,32 @@ def __init__(self, start: Number, step: Number, length: int, dtype: DType = None
             >>> import netket as nk
             >>> n_max = 10
             >>> nk.utils.StaticRange(start=-1, step=2, length=2)
-            StaticRange(start=-1, step=2, length=2, dtype=int64)
+            StaticRange(start=-1, step=2, end=1, length=2, dtype=int64)
 
         Args:
             start: Value of the first entry
             step: Step between the entries
             length: Length of this range
             dtype: The data type
         """
-        dtype = canonicalize_dtypes(start, step, dtype=dtype)
+        end = start + step * (length - 1)
+
+        if dtype is None:
+            dtype = canonicalize_dtypes(start, step, dtype=dtype)
```

Comment:
You could add something like
```suggestion
            dtype = canonicalize_dtypes(start, step, dtype=dtype)
            dtype = bottom_dtypes(start, step, length, dtype=dtype)
```

and define bottom types to be something like
```
In [46]: def bottom_dtypes(start, step, length, dtype):
    ...:     if np.issubdtype(dtype, np.floating):
    ...:         return dtype
    ...:     int_dtypes = [np.int8, np.int16, np.int32, np.int64]
    ...:     uint_dtypes = [np.uint8, np.uint16, np.uint32, np.uint64]
    ...:     end = start + step*(length-1)
    ...:
    ...:     dtypes_choice = uint_dtypes
    ...:     if any(v<0 for v in (start, end)):
    ...:         dtypes_choice = int_dtypes
    ...:     elif start == 0 and length==2 and end==1:
    ...:         return np.dtype(bool)
    ...:
    ...:
    ...:     for dtyp in dtypes_choice:
    ...:         print(f"testing {dtyp}")
    ...:         if all(_in_int_dtype_range(v, dtyp) for v in (start, end)):
    ...:             return dtyp
    ...:     raise ValueError("This should never happen")
```

which should give the smallest dtype to contain start and stop.

Reply:
Currently, NetKet has the default type int/float (perhaps `np.int64` and `np.float64`). 
I intended to first fix the bug (without changing tests), and then change the behavior separately, perhaps with #1938.

When I added a similar function, [L66 or L67](https://github.com/netket/netket/blob/e0a523f8742bc1cee247d6bd262303266d94b785/test/utils/test_range.py#L66C5-L67C57) in the test_range.py failed. If I return a Python's scalar it doesn't have a dtype, if I return an array it is the minimal integer and not `int`.
Additionally, it is not clear to me when NetKet expects Python's scalar (i.e., 64-bit floats and integers), scalars in an array or a NumPy scalar (`np.int8(x)`).

Reply:
> Additionally, it is not clear to me when NetKet expects Python's scalar (i.e., 64-bit floats and integers), scalars in an array or a NumPy scalar (np.int8(x)).

It's not well defined at all. If tests pass, assume the code is ok.
The code now was written to work in all cases, but maybe it's an overkill.